### PR TITLE
[Ready] Documentation and setup for blog

### PR DESCRIPTION
**Pull request by enzoevers**

### CODE REVIEW

**Comment by enzoevers** [comment]
It's this document: https://github.com/Byont-Ventures/smart-contract-analysis-tools/blob/0d93ac01a965978df62fd32c81c442219fac4173/docs/analysis-techniques.md

File: docs/analysis-techniques.md [hunk 1]
```diff
@@ -9,31 +9,31 @@
...
-/// @notice Paying off existing dept, or adding credit.
+/// @notice Paying off existing dept or adding credit.
 /// @param amount The amount to add to the credit.
 function payOff(uint256 amount) public {
     credit[msg.sender] += int256(amount);
 }
 ```
 
-On first sight it seems just fine. However, not that an `uint256` is casted to an `int256`. Solidity uses [two's complement](https://nl.wikipedia.org/wiki/Two%27s_complement) for representing negative values. Meaning that where `uint256` has the range [0, $2^{256} - 1$], `int256` has the range [$-2^{255}$, $-2^{255} - 1$].
+At first sight, it seems just fine. However, note that an `uint256` is cast to an `int256`. Solidity uses [two's complement](https://nl.wikipedia.org/wiki/Two%27s_complement) for representing negative values. Meaning that where `uint256` has the range [0, $2^{256} - 1$], `int256` has the range [$-2^{255}$, $-2^{255} - 1$].
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:36]: a uint (long u is treated as a consonant)

Comment (SebastiaanCrisan) [reply docs/analysis-techniques.md:36]: [$-2^{255}$, $-2^{255} - 1$]
$-2^{255}$ is not getting formatted for some reason

File: docs/analysis-techniques.md [hunk 2]
```diff
@@ -43,29 +43,29 @@ If this function could be naively tested, a test could be created to only check
...
 
 What you will notice when diving more into these different (non-)automated tooling is that there will always be a bit of both, while there are clear differences between techniques and tooling, there is also quite some overlap.
 
-One of the techniques used as the backbone of both non-automated and automated technique is [Satisfiable Modulo Theory (SMT)](#satisfiable-modulo-theory-smt). The main purpose of SMT is to check if the variables in a program can have a certain (initial) value such that a requirement is met. In other words, if there is a **satisfiable** assignment for the variables.
+One of the techniques used as the backbone of both non-automated and automated techniques is [Satisfiable Modulo Theory (SMT)](#satisfiable-modulo-theory-smt). The main purpose of SMT is to check if the variables in a program can have a certain (initial) value such that a requirement is met. In other words, if there is a **satisfiable** assignment for the variables.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:66]: [Satisfiable Modulo Theory (SMT)](#satisfiable-modulo-theory-smt) link does not seem to work

File: docs/analysis-techniques.md [hunk 3]
```diff
@@ -43,29 +43,29 @@ If this function could be naively tested, a test could be created to only check
...
 
-A technique that uses SMT is [Symbolic execution](#symbolic-execution) that checks all branches of a program to see if any of them lead to a failing assertions. If it finds a failing branch it will check if that branch can be reached using SMT. But SMT is also used during other stages in symbolic execution. For example to avoid wasting processing time on branches that can't be reached anyway. In other words to prune (remove) these branches from the analysis process.
+A technique that uses SMT is [Symbolic execution](#symbolic-execution) which checks all branches of a program to see if any of them lead to a failing assertion. If it finds a failing branch it will check if that branch can be reached using SMT. But SMT is also used during other stages in symbolic execution. For example to avoid wasting processing time on branches that can't be reached anyway. In other words to prune (remove) these branches from the analysis process.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:68]: [Symbolic execution](#symbolic-execution) link does not seem to work

File: docs/analysis-techniques.md [hunk 4]
```diff
@@ -221,61 +221,61 @@ flowchart TB
...
 
-The main goal of static analysis tools is to find patterns in the code that are not conforming to a certain standard, or that are known to have a high chance of contributing to a bug in the code. Another thing is that is could, for example, find dead code by looking where function and variables are used in the code.
+The main goal of static analysis tools is to find patterns in the code that are not conforming to a certain standard, or that are known to have a high chance of contributing to a bug in the code. Another thing is that it could, for example, find dead code by looking at where functions and variables are used in the code.
 
 For Solidity one such pattern is for example updating a variable, used to check if an external call has to be called, only after the external call. This is a classic setup for re-entrancy attacks.
 
 Note that [Prettier](https://prettier.io/) (the code formatter) is also a static analysis tool. It looks at your code and finds violations of rules defined in the configuration of prettier.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:250]: Nice!

File: docs/analysis-techniques.md [hunk 5]
```diff
@@ -221,61 +221,61 @@ flowchart TB
...
 
-Also, the tools can only verify what is given to them, so if the assertion does not cover the associated required in a function design, then the assertion might not be violated while the assertion itself is not correct. Giving a false impression of success.
+Also, the tools can only verify what is given to them, so if the assertion does not cover the associated requirement in a function design, then the assertion might not be violated while the assertion itself is not correct. Giving a false impression of success.
 
 ---
 
 ## 9 How Byont uses these techniques
 
-At Byont we make heavy use of Foundry's fuzzing capabilities. Besides that, we are also making use of Slither, solc's SMTChecker, Mythril and KEVM. We are currently in the process of optimizing the process of incorporating these tools in our development flow.
+At Byont we make heavy use of Foundry's fuzzing capabilities. Besides that, we are also making use of Slither, solc's SMTChecker, Mythril, and KEVM. We are currently in the process of optimizing the process of incorporating these tools in our development flow.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:264]: Might be nice to explain a bit more about what KEVM is / does or provide a link to a resource that does so

**Review by SebastiaanCrisan — CHANGES_REQUESTED** [review]

File: docs/analysis-techniques.md (hunk 3 above)
Comment (enzoevers) [reply docs/analysis-techniques.md]: [afac119](https://github.com/Byont-Ventures/smart-contract-analysis-tools/pull/47/commits/afac1198df618d42b4f78a0b408d8372ecdcde8a)

File: docs/analysis-techniques.md (hunk 2 above)
Comment (enzoevers) [reply docs/analysis-techniques.md]: [afac119](https://github.com/Byont-Ventures/smart-contract-analysis-tools/pull/47/commits/afac1198df618d42b4f78a0b408d8372ecdcde8a)

File: docs/analysis-techniques.md (hunk 1 above)
Comment (enzoevers) [reply docs/analysis-techniques.md]: Currently just using plain text to avoid rendering errors in other viewers: [afac119](https://github.com/Byont-Ventures/smart-contract-analysis-tools/pull/47/commits/afac1198df618d42b4f78a0b408d8372ecdcde8a)

File: docs/analysis-techniques.md [hunk 6]
```diff
@@ -9,31 +9,31 @@
...
-Every developer writes unit-tests for there code (right?). The goal of such a unit test is to test that a single function does what it is expected to do. Or, depending on the usage, testing that a certain function calls other function to perform a local integration test. Sounds easy enough.
+Every developer writes unit tests for their code (right?). The goal of such a unit test is to test that a single function does what it is expected to do. Or, depending on the usage, testing that a certain function calls another function to perform a local integration test. Sounds easy enough.
 
 Let's have a simple function as seen below:
 
 ```solidity
 mapping(address => int256) credit;
 
-/// @notice Paying off existing dept, or adding credit.
+/// @notice Paying off existing dept or adding credit.
```
[the remark this answers is not in the document]
Comment (enzoevers) [reply docs/analysis-techniques.md]: I knew that

File: docs/analysis-techniques.md (hunk 5 above)
Comment (enzoevers) [reply docs/analysis-techniques.md]: [da3b96a](https://github.com/Byont-Ventures/smart-contract-analysis-tools/pull/47/commits/da3b96aa5950dc586ca73112f0d73e6b6a8fb1c4): I've update this part a little bit by adding links to the docs and git repo's of the different tools. I think that this page should not go into much more details about the tools themselves. Both because 1) it can become quite long what going into the differences of each tools, and 2) because in a next blog/documentation page, the different tools will be demonstrated by means of a demo.

**Comment by jasperverbeet** [comment]
I asked chat gtp to write an introduction to the analysis techniques, here are the 2 options:

```
This article discusses various analysis techniques that can be used to ensure the quality and correctness of code. The techniques covered include unit testing, fuzzing (property testing), automated testing, Satisfiable Modulo Theory (SMT), symbolic execution, static analysis, and formal verification. The article also discusses how the Byont project utilizes these techniques. Each technique is described in detail, including examples and any limitations or considerations. The article concludes with a list of additional sources for further reading on the topic.
```

and 

```
This article covers various techniques for analyzing and testing code, including unit testing, fuzzing, automated testing, Satisfiable Modulo Theory (SMT), symbolic execution, static analysis, and formal verification. We'll go over each technique in detail and provide examples and considerations for each. We'll also discuss how the Byont project uses these techniques. Whether you're a seasoned developer or just starting out, there's something in this article for you. Get ready to take your code to the next level!
```

**Comment by Vitality-Booster** [comment]
1) Analysis techniques of what? The context a little unclear when I began reading the file. It can be understood from the repository it is in, but I would suggest adding small intro for the document, 2-3 lines would be enough. If you want to make it a blog, then definitely make an introduction.

2) A small tip: it is better to avoid pronouns. It's difficult, but that way the article sounds more professional in a sense.

3) Fuzzing: 
	a) "So now you really need to think of what the behavior is instead of what the result should be. " sounds a bit unclear and "complex". Suggestion: "This kind of test is aimed more at checking the behavior of the function, rather than its output". Something like that.
	b) "Fuzzing, however, still is only running unit tests." not clear what it means, seems like in this context "unit tests" are seen as something bad/inefficient (?), but I don't why, I think it'd be better to include some more context there.

4) Can we do better?
	a) Also to make the article look more professional, try to avoid "interacting with a reader". Example there: "Great!" and "Not so great!". I would suggest trying to avoid that.

5) Automated testing
	a) Check everything with Grammarly, as there are some typos here and there. Example: "This can either the be actual source code ... ."
	b) "For these techniques, the user has to define a model of how to design works (the design, not the code)." Design works? I didn't understand that, could you elaborate on that?
	c) Could you in the second paragraph, maybe, reference an article that shows a difference between automated and non-automated verifications? Or maybe elaborate on that more, because I'm still a bit confused how it works with this "design" thingy that I'm also confused with. 
	d) "... the model and the properties and check if they can be satisfied" what exactly can be satisfied there? I'm confused there as well, honestly.
	e) The part about SMT and Symbolic execution in this chapter sounds pretty good to me.

6) Satisfiable Modulo Theory (SMT)
	a) "Note that the assumptions and requirements can originate from knowledge about the system in which the function is used, while the function itself doesn't have this knowledge explicitly. The function might be fine when used in the system, but not when used as a self-containing function." I think I understand what you mean there, but I'm not sure that I understood why you included that. Try to explain the reason you mentioned that. Example: "So, it is better to write self-containing functions for SMT". Or something like that.
	b) "if the requirement (a + b > 100) can hold at al." hold what? Or maybe that some sort phrase that I don't know. But it is unclear for me.
	c) "Running this in the online tool results in the following output. Saying that if a = 16 and b = 85 that all constraints are met (which is true since 16 + 85 = 101 > 100)." Do you mean in the first sentence that the code below is the output? If yes, then I would after the first sentence put the code and then beneath the code mention: "This code means that f a = 16 and b = 85 that all constraints are met (which is true since 16 + 85 = 101 > 100)."
	d) The question for the whole z3. Does it mean that in the output it will only tell if the last "assert" is resulting in something or not? Or maybe, with every assert it limits the values that it can use, until there are no more values that satisfies all the asserts and then it sends "unsat" or if there are no more asserts that limit the values it just shows the first situation where all the requirements are met? Maybe, you could explain this flow a bit more in depth. (UPD: was kind of explained in 6.1, but maybe it would be good to also mention here that this question is explained further).

7) Symbolic execution
	a) "One of the things that are done to reduce the state-space and resource usage is to prune uninteresting states. Pruning was briefly mentioned in the symbolic execution example." Add a link to a good article there or maybe try to explain it yourself, but don't leave just like that.
	b) In 6.2 in the 3rd paragraph you mention "state-space-explosion", then in the 4th one you tell about "pruning" and in the 5th you return back to the "state-space-explosion". Firstly, finish one topic, then switch to the other one.
	c) The last paragraph in 6.6. 
		i) Firstly, it is not clear why the tool could return a random value or give up (is it due to the tool or because of the fact that we don't know the function code).
		ii) Secondly, it is not clear why you mentioned that. It looks more like a random fact. I would suggest in 6.2 say in the beginning: "The example we used is very simple, but what if a more difficult function is tested with Symbolic execution? There are some limits to what you can do: /n 1. Recursive functions and loops. ... /n 2. [something here] /n ... /n 4. External functions. ...". Something like that.

8) Static analysis
	a) "(be it with symbolic values)" sounds a bit weird and a bit unclear.
	b) "contributing to a bug in the code" also sounds a bit weird. Maybe, better "leading to / resulting into a bug".
	c) "check if an external call has to be called, only after the external call" I guess you meant "only after the value is updated".
	d) "This is because static analysis can be used to get a better picture of the whole code with all its dependencies." why, is it taken for granted? Better to have a link that explains that or your own explanation. I think that this part "Meaning that branches that are not interacting with the state variables can be pruned earlier for example." doesn't explain why static analysis gives a broader picture of the code.

More general tips:

	1) Try to sound more professional. Avoid pronouns and "interaction with a reader".
	2) Sometimes your article looks like a bunch of facts that are not connected to each other and because of that sometimes I (as a reader) do not understand what is the point of some information or what you are trying to say. Try to make the article fluently and logically go from one point to another.
	3) Sometimes also you don't explain enough. You just briefly mention something and move to the other topic. If you just don't want this article to focus on these parts, but you think they are worth mentioning, then just give a source link where people could learn more about the topic. Or you can try to explain yourself, but it is hard to explain some aspect, when you don't want to focus on it too much.
	4) Sometimes your language is also hard to understand. I don't know how to help with, just ask for feedback and learn by trial-and-error.

Note: things I mention in general tips, I sometimes pointed out to it earlier, but not everywhere, so you will need to find out some stuff on your own.

Hopefully, that doesn't sound too harsh and nice job for trying writing an article :)

**Comment by jasperverbeet** [comment]
@enzoevers @Vitality-Booster both very well done! Enzo very nice, interesting and on-topic article. And Vitali thanks for your amazing feedback. 💪🏼💪🏼 keep it up guys.

**Comment by jasperverbeet** [comment]
Also @enzoevers you're allowed to give yourself credit in the introduction by telling that you've researched and worked with all these tools. If you choose not  to mention it we'll probably do it when posting in on our socials anywhere 🙌🏼. Up to you...

**Comment by enzoevers** [comment]
@Vitality-Booster Thanks for the detailed feedback. It's really useful.

**Review by SebastiaanCrisan — APPROVED** [review]

File: docs/analysis-techniques.md [hunk 7]
```diff
@@ -1,21 +1,24 @@
...
-  - [8 A note of formal verification](#8-a-note-of-formal-verification)
-  - [9 How Byont uses these techniques](#9-how-byont-uses-these-techniques)
-  - [10 Follow-up](#10-follow-up)
-  - [More sources](#more-sources)
+This article describes several techniques such as fuzzing (property testing), symbolic execution, static analysis, and Satisfiability Modulo Theory (SMT) that can be used as an addition to the more standard unit testing. It is also described these techniques are working together. Each technique is explained by means of examples. The limitations and the therefrom following considerations are described as well. At the end of the article it is described how Byont utilizes these techniques for their report generation.
+
+The layout is as follows:
+
+- [1 Unit testing](#1-unit-testing)
+- [2 Fuzzing (property testing)](#2-fuzzing-property-testing)
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:10]: this one is not working for me

**Review by SebastiaanCrisan — CHANGES_REQUESTED** [review]

**Review by SebastiaanCrisan — APPROVED** [review]

**Review by SebastiaanCrisan — APPROVED** [review]

**Review by SebastiaanCrisan — APPROVED** [review]

File: docs/analysis-techniques.md [hunk 8]
```diff
@@ -24,9 +24,9 @@ The layout is as follows:
 
 ## 1 Unit testing
 
-Every developer writes unit tests for their code (right?). The goal of such a unit test is to test that a single function does what it is expected to do. Or, depending on the usage, testing that a certain function calls another function to perform a local integration test. Sounds easy enough.
+The goal of a unit tests is to test that a function does what it is expected to do.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:27]: small typo: unit tests => unit test

File: docs/analysis-techniques.md [hunk 9]
```diff
@@ -36,23 +36,25 @@ mapping(address => int256) credit;
...
+This is all to say that unit testing is good, as long as the developer created unit tests for all edge cases.
 
 ## 2 Fuzzing (property testing)
 
-Another technique is to use fuzzing on your unit tests. Fuzzing is also called property testing. This is because instead of testing a single scenario, your test now needs to work for **all** possible values (in the range of the type of course). So now you really need to think of what the behavior is instead of what the result should be. The [Foundry](https://book.getfoundry.sh/forge/fuzz-testing?highlight=fuzz#fuzz-testing) framework has this built-in.
+Another technique is to use fuzzing (on unit tests). Fuzzing is also called property testing. This is because instead of testing a single scenario, your test now needs to work for **all** possible values (in the range of the type of course). in other words, the test should focus on the behavior of the function rather than the output. The [Foundry](https://book.getfoundry.sh/forge/fuzz-testing?highlight=fuzz#fuzz-testing) framework for smart-contracts has this built-in.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md]: > (in the range of the type of course). in other words,

in other words => In other words

**Review by SebastiaanCrisan — CHANGES_REQUESTED** [review]

**Review by SebastiaanCrisan — APPROVED** [review]

**Review by SebastiaanCrisan — APPROVED** [review]

File: test.js [hunk 10]
```diff
@@ -0,0 +1 @@
+function calculateDaysBetweenDates(begin, end) {
```
Comment (SebastiaanCrisan) [inline test.js:1]: I think you can remove this file(?)

**Review by SebastiaanCrisan — CHANGES_REQUESTED** [review]

File: test.js (hunk 10 above)
Comment (SebastiaanCrisan) [reply test.js:1]: Rest is looking good 👍

Comment (enzoevers) [reply test.js:1]: Jep was committed when testing github copilot.

**Comment by enzoevers** [comment]
@SebastiaanCrisan I see that there are quite some consecutive approves from you. Is this because you are looking at each commit individually? If so, I think it is best to look at the changes of a couple of commits together after some time to save doing double work and to see the changes in the bigger picture.

Showing only a selection of commits can be done when being in the "Files changed" tab and then going to the "Changes from ..." drop-down menu that is just above the list with all the files.

Just saying this as a tip. But thanks for the help with the reviews c:

File: docs/analysis-techniques.md [hunk 11]
```diff
@@ -45,76 +48,144 @@ function payOff(uint256 amount) public {
 ...
 ```
 
-At first sight, it seems just fine. However, note that a `uint256` is cast to an `int256`. Solidity uses [two's complement](https://nl.wikipedia.org/wiki/Two%27s_complement) for representing negative values. Meaning that where `uint256` has the range `[0, 2^256 - 1]`, `int256` has the range `[-2^255, -2^255 - 1]`.
+At first sight, it seems just fine. However, note that a `uint256` is casted to an `int256`. Solidity uses [two's complement](https://nl.wikipedia.org/wiki/Two%27s_complement) for representing signed values. Meaning that where `uint256` has the range `[0, 2^256 - 1]`, `int256` has the range `[-2^255, -2^255 - 1]`.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:51]: Cast instead of casted was fine!

File: docs/analysis-techniques.md [hunk 12]
```diff
@@ -281,70 +371,64 @@ function getAgeCategory(uint256[] memory age) public pure returns (AGE_CATAGORY[
...
+So far SMT, symbolic execution and fuzzing have been discussed as tools that help to find more bugs than with unit tests alone. These are all analysis tools that check the source code against assertions/requirements that the code should adhere to by running the code. Static analysis is different in that it doesn't run the code, but it looks at the code and tries to find patterns that are known to be problematic.
 
-In another [paper](https://www.fit.vutbr.cz/research/groups/verifit/tools/muse/bytecode08.pdf), symbolic execution was used to generate a state-machine on which temporal properties could be verified. This shown again that verification techniques don't exclude each other.
+For example, [Prettier](https://prettier.io/) (the code formatter) is also static analysis tool. It looks at your code and finds violations of rules defined in the configuration of prettier. Also [Grammarly](https://app.grammarly.com/) can be though of as a static analysis tool. It looks at your text and finds violations of rules defined by the English language.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:389]: Is also static analysis tool => Is also a static analysis tool 
Can be though of => Can be thought of

File: docs/analysis-techniques.md [hunk 13]
```diff
@@ -281,70 +371,64 @@ function getAgeCategory(uint256[] memory age) public pure returns (AGE_CATAGORY[
...
 - solc's [SMTChecker](https://docs.soliditylang.org/en/v0.8.17/smtchecker.html) and [Mythril](https://github.com/ConsenSys/mythril/tree/develop) (symbolic execution)
 - [KEVM](https://github.com/runtimeverification/evm-semantics) (symbolic execution and more).
 
-KEVM is built on the [K-framework](https://github.com/runtimeverification/k) from [Runtime Verification](https://runtimeverification.com/).
-
-We are currently in the process of improving the integration of these tools in our development flow.
+How these tools work in detail will be discussed in a later article.
 
-It is important to use multiple tools since none of them will find all problems on their own. This is not because the tools are not good enough (because they are good). It's just that their methods of implementing the analysis techniques can focus on different aspects.
+Byont is currently in the process of improving the integration of these tools in it's smart contract development flow.
```
Comment (SebastiaanCrisan) [inline docs/analysis-techniques.md:417]: it's => its

**Review by SebastiaanCrisan — CHANGES_REQUESTED** [review]
